### ISSUE TITLE
Support OpenType math

### ISSUE THREAD

**Issue by GoogleCodeExporter** [notofonts/math#14]
```
If Noto supported OpenType MATH tables, the font could be used for rendering 
mathematical equations in high quality. In particular, (a) web browsers could 
use the Noto font for rendering MathML; and (b) TeX could use Noto for 
high-quality mathematical typesetting.

Some pointers:
http://fontforge.github.io/math.html
http://www.tug.org/tugboat/tb30-1/tb94vieth.pdf
https://tug.org/TUGboat/tb33-3/tb105vieth.pdf
http://blogs.msdn.com/b/murrays/archive/2014/04/27/opentype-math-tables.aspx
http://mpeg.chiariglione.org/standards/mpeg-4/open-font-format/text-isoiec-cd-14
496-22-3rd-edition
https://wiki.mozilla.org/MathML:Open_Type_MATH_Table
https://developer.mozilla.org/en-US/docs/Mozilla/MathML_Project/Fonts#MATH_fonts

```

Original issue reported on code.google.com by `sas...@google.com` on 8 Apr 2015 at 8:19

**Comment by apoorvpotnis** [comment]
> This is the summary of the discussion we had this morning at the web engines hackfest:
> 
>     * Introducing an OpenType MATH table to Noto Sans Math is an option, but that will likely require some effort. An alternative would be to ship available open source fonts that already exist.
> 
>     * One option is to use STIX Two Math under the SIL Open Font License ( https://www.stixfonts.org/ ). STIXTwoMath-Regular.otf is 819KB and STIXTwoMath-Regular.woff2 is 539KB. On the one hand perhaps this is too large (e.g. we know Android has very strict limitation about not increasing chromium binary size) but on the othe hand Android already ships large fonts?
> 
>     * One idea is to subset the STIX Two Math fonts to only keep the most important parts (e.g. MATH table, stretchy constructions, etc). This was experimented some years ago at https://github.com/fred-wang/MOZTIX where the OTF was 152KB and the WOFF2 99KB.
> 
>     * Another option is Latin Modern Math, which is closer to the look & feel of the default font used by TeX: https://www.gust.org.pl/projects/e-foundry/lm-math

New Computer Modern (Book weight) is an available alternative as well, since it has glyphs not only for all the unicode math blocks but additional symbols as well. The book weight is thicker than Latin Modern, although not as thick as the current MathJax font. CTAN: https://ctan.org/pkg/newcomputermodern

**Comment by apoorvpotnis** [comment]
I am not exactly sure if this is the proper way to check this, but what I tried was I ran `./ots-sanitize NewCMMath-Book.otf` using [this repository](https://github.com/khaledhosny/ots). I got the message `File sanitized successfully!`. I tried to run `ots-validator-checker` but `catchsegv` and `valgrind` don't seem to work.

**Comment by apoorvpotnis** [comment]
> @apoorvpotnis Does New Computer Modern Math pass OTS validation?

@fred-wang 
I am not exactly sure if this is the proper way to check this, but what I tried was I ran `./ots-sanitize NewCMMath-Book.otf` using [this repository](https://github.com/khaledhosny/ots). I got the message File sanitized successfully!. I tried to run `ots-validator-checker` but `catchsegv` and `valgrind` don't seem to work.

**Comment by simoncozens** [comment]
This is now fixed, hooray!

**Comment by fred-wang** [comment]
@simoncozens: Do you have a link to a Noto Math font we can try?
The ones at https://notofonts.github.io/math/ seem either to lack a MATH table, or to have completely wrong. glyph mappings in the MathVariants subtable. The glyph names in NotoSansMath-Regular-MATH-table.ttx looks ok, though, so maybe something is happening during the conversion?

**Comment by simoncozens** [comment]
Thanks for testing this. The ones with the MATH table (obviously!) *should* work, so something must have gone wrong when adding the ttx dump to the binary. I'll take a look. I'll also add MATH tables to the "full" and "googlefonts" builds.

**Comment by simoncozens** [comment]
Oh, heck, I know what the problem is.

**Comment by simoncozens** [comment]
For reference, here's what the problem was: the TTX file we were sent had coverage tables which listed the glyphs in a different glyph order to the one produced by fontmake, so we were getting `GSUB/GPOS Coverage is not sorted by glyph ids` warnings and it was falling fontbakery and otsanitize and everything was bad.

So, *completely failing to remember how coverage tables worked*, I hand-hacked the TTX to reorder the coverage tables - but I did *not* reorder the things that the coverage tables pointed to. Oops!

Now I have written a script which applies the TTX and then re-orders both the coverage tables and their referent array elements in the same order. Please try the latest development builds. If they work, I'll release a new version.

**Comment by simoncozens** [comment]
Grr, extensible symbols don't seem to be working:

<img width="570" alt="Screenshot 2023-03-15 at 12 06 00" src="https://user-images.githubusercontent.com/106728/225304120-cd48da39-1b34-4690-bf6b-65566269ee85.png">

**Comment by apoorvpotnis** [comment]
> Grr, extensible symbols don't seem to be working:
> 
> <img alt="Screenshot 2023-03-15 at 12 06 00" width="570" src="https://user-images.githubusercontent.com/106728/225304120-cd48da39-1b34-4690-bf6b-65566269ee85.png">

Is this how Noto _**Sans**_ math is supposed to look, modulo extensible delimiters? Shouldn't the symbols be in sans?

**Comment by simoncozens** [comment]
Urgh. I think the font is actually doing the right thing here; there are glyphs for:

* U+1D5BA (MATHEMATICAL SANS-SERIF SMALL A)
* U+1D622 (MATHEMATICAL SANS-SERIF ITALIC SMALL A)
* U+1D44E (MATHEMATICAL ITALIC SMALL A)

![shape](https://user-images.githubusercontent.com/106728/225317809-8ba82af7-0f75-4236-86ae-2ad1ca088371.png)

But LaTeX is kind of assuming serif-ness and mapping `$a$` to U+1D44E rather than U+1D622. I think it would be wrong to make U+1D622 and U+1D44E point to the same glyph, which I why I think the font is fine.

I don't know why the delimiters aren't working. The MATH table looks OK to me...

**Comment by apoorvpotnis** [comment]
I am not a LaTeX or font expert, but here is what I think. If the math font were serif, then I would expect $a$ to give me U+1D44E (serif) and if the math font was supposed to be sans, then I would expect $a$ to give me a sans serif _a_. Perhaps this should be implemented on the LaTeX level and not on the font level as you say.

**Comment by apoorvpotnis** [comment]
In LaTeX, one could define all the math letters to map to the corresponding sans serif glyph, but the unicode standard does not have italic sans serif greek. So, there is nothing to map italic greek symbols.

**Comment by simoncozens** [comment]
> I don't know why the delimiters aren't working. The MATH table looks OK to me...

The MATH table is correct. The problem is that these maths glyphs are implemented using Glyphs smart components.

**Comment by simoncozens** [comment]
Fixed again, with new version of glyphsLib.

**Comment by fred-wang** [comment]
@simoncozens Thanks! From a quick test, this seems to work now. This is how a [Wikipedia](https://en.wikipedia.org/wiki/Cauchy%E2%80%93Schwarz_inequality) page renders with Noto Sans in Chromium 111:

![wikipedia-chromium-noto-sans](https://user-images.githubusercontent.com/567455/225527011-6814722e-e0a3-41ae-a99b-c1c13f8b767c.png)

**Comment by davidcarlisle** [comment]
> Urgh. I think the font is actually doing the right thing here; there are glyphs for:
> 
>     * U+1D5BA (MATHEMATICAL SANS-SERIF SMALL A)
> 
>     * U+1D622 (MATHEMATICAL SANS-SERIF ITALIC SMALL A)
> 
>     * U+1D44E (MATHEMATICAL ITALIC SMALL A)
> 
> 
> ![shape](https://user-images.githubusercontent.com/106728/225317809-8ba82af7-0f75-4236-86ae-2ad1ca088371.png)
> 
> But LaTeX is kind of assuming serif-ness and mapping `$a$` to U+1D44E rather than U+1D622. I think it would be wrong to make U+1D622 and U+1D44E point to the same glyph, which I why I think the font is fine.
> 
> I don't know why the delimiters aren't working. The MATH table looks OK to me...

I must say I was very surprised by the serifs here when I tried the font.

![image](https://user-images.githubusercontent.com/1268738/226957980-1ec548ab-2aea-4856-ab1f-8030cc59104c.png)


Compared to selecting Fira Math, as an existing sans serif math font

![image](https://user-images.githubusercontent.com/1268738/226958811-5e1f1b84-4bbc-431e-893d-249e84b51297.png)


That was lualatex, but in mathml it would be the same `<mi>a</mi>` is by definition U+1D44E  if then globally you chose to set the document in sans serif it would be good to just specify the font to use not have to edit the individual characters, which may have been entered as `<mi>a</mi>` or  `<mi>𝑎</mi>` or ...

In tex it's not really much of an issue as the mapping can be adjusted on the fly but for mathml and similar systems would it be more natural to have the default style for letters to be sans serif, at least as an option?

**Comment by khaledhosny** [comment]
> I must say I was very surprised by the serifs here when I tried the font.

FYI, this (and many other inconsistencies), should be fixed with #57.

**Comment by fred-wang** [comment]
FYI, I opened https://issuetracker.google.com/issues/498266576 to ship Noto Sans Math in Android.
